### PR TITLE
Feature/directory caching

### PR DESCRIPTION
**Pull request by gnusi**

### CODE REVIEW

File: core/store/caching_directory.hpp [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    if (!stream) {
+      return nullptr;
+    }
+
+    {
+      std::lock_guard lock{mutex_};
+      if (Acceptor::operator()(cache_.size(), name, advice)) {
+        try {
+          const auto [it, _] = cache_.try_emplace(name, std::move(stream));
+          return it->second->reopen();
```
Comment (MBkkt) [inline core/store/caching_directory.hpp:123]: same here

File: core/store/caching_directory.hpp [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        } catch (...) {
+        }
+      }
+      return true;
+    }
+    return false;
+  }
+
+  index_input::ptr open(std::string_view name,
+                        IOAdvice advice) const noexcept override {
```
Comment (MBkkt) [inline core/store/caching_directory.hpp:103]: I think better to write separate OpenImpl function which return index_input::ptr and here you just call reopen on it (if not null)

File: core/store/caching_directory.hpp [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+  bool rename(std::string_view src, std::string_view dst) noexcept override {
+    if (Impl::rename(src, dst)) {
+      std::lock_guard lock{mutex_};
+      if (auto src_it = cache_.find(src); src_it != cache_.end()) {
+        auto tmp = std::move(src_it->second);
+        cache_.erase(src_it);
+        try {
+          cache_[dst] = std::move(tmp);
+        } catch (...) {
+        }
```
Comment (MBkkt) [inline core/store/caching_directory.hpp:95]: it's not ok if we ignore exception here, I think if we don't have enough memory in this situation it's ok to terminate

File: core/store/mmap_directory.hpp [hunk 4]
```diff
@@ -32,10 +31,7 @@ class mmap_directory : public fs_directory {
   explicit mmap_directory(std::filesystem::path dir,
                           directory_attributes attrs = directory_attributes{});
 
-  virtual index_input::ptr open(std::string_view name,
-                                IOAdvice advice) const noexcept override final;
+  index_input::ptr open(std::string_view name, IOAdvice advice) const noexcept;
```
Comment (MBkkt) [inline core/store/mmap_directory.hpp:34]: is it virtual? override?

File: core/store/caching_directory.hpp [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+};
+
+template<typename Impl, typename Acceptor = MaxSizeAcceptor>
+class CachingDirectory : public Impl, private Acceptor {
+ public:
+  template<typename... Args>
+  explicit CachingDirectory(const Acceptor& acceptor, Args&&... args)
+    : Impl{std::forward<Args>(args)...}, Acceptor{acceptor} {}
+
+  bool exists(bool& result, std::string_view name) const noexcept override {
```
Comment (MBkkt) [inline core/store/caching_directory.hpp:56]: I think all member function, at least now can be final

File: core/store/caching_directory.hpp [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+
+    auto stream = Impl::open(name, advice);
+
+    if (!stream) {
+      return nullptr;
+    }
+
+    {
+      std::lock_guard lock{mutex_};
+      if (Acceptor::operator()(cache_.size(), name, advice)) {
```
Comment (MBkkt) [inline core/store/caching_directory.hpp:120]: what is acceptor?
What if two threads create "same" stream, wait mutex,
For first thread acceptor return true -> make reopen
For second false -> return stream?

File: core/store/caching_directory.hpp [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+        try {
+          const auto [it, _] = cache_.try_emplace(name, std::move(stream));
+          return it->second->reopen();
+        } catch (...) {
+        }
+      }
+    }
+
+    return stream;
```
Comment (MBkkt) [inline core/store/caching_directory.hpp:129]: why?

in general I expected something like

```cpp
compute hash;
if (shared_lock lock{mutex}; auto input = get(key, hash)) {
  lock.unlock();
  return input.reopen();
}
auto input = Impl::open(name, advice);
unique_lock lock{mutex};
const auto [it, was_insert] = cache_.try_emplace(name, std::move(input));
if (was_insert) {
  lock.unlock();
  return input;
} else {
  auto old_input = it->second;
  lock.unlock();
  input.close/remove();
  return old_input->reopen();
}
```

File: core/store/caching_directory.hpp (hunk 1 above)
Comment (MBkkt) [inline core/store/caching_directory.hpp:123]: You call reopen for file which was open few lines ago

**Review by MBkkt — CHANGES_REQUESTED** [review]

File: core/store/caching_directory.hpp [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+
+#include <shared_mutex>
+
+#include "store/directory.hpp"
+
+namespace iresearch {
+
+class MaxSizeAcceptor {
+ public:
+  explicit constexpr MaxSizeAcceptor(size_t max_size = 1024) noexcept
```
Comment (Dronplane) [inline core/store/caching_directory.hpp:36]: NIT 
MaxCountAcceptor would be better. As it is count of files, not overall size

File: core/store/caching_directory.hpp [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+        try {
+          cache_[dst] = std::move(tmp);
+        } catch (...) {
+        }
+      }
+      return true;
+    }
+    return false;
+  }
+
```
Comment (MBkkt) [inline core/store/caching_directory.hpp:101]:
```cpp
bool find(key, Visitor&& v) {
      hash = std::hash<>{}(key);
      std::shared_lock lock{mutex_};
      if (auto it = cache_.find(name, hash); it != cache_.end()) {
          v(it);
          return true;
      }
      return false;
}
```

**Review by Dronplane — APPROVED** [review]
LGTM

File: core/store/caching_directory.hpp [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+};
+
+template<typename Impl, typename Acceptor = MaxCountAcceptor>
+class CachingDirectory : public Impl, private Acceptor {
+ public:
+  template<typename... Args>
+  explicit CachingDirectory(const Acceptor& acceptor, Args&&... args)
+    : Impl{std::forward<Args>(args)...}, Acceptor{acceptor} {}
+
+  bool exists(bool& result, std::string_view name) const noexcept override {
```
Comment (MBkkt) [inline core/store/caching_directory.hpp:56]: final?

**Review by MBkkt — APPROVED** [review]

File: core/store/caching_directory.hpp [hunk 11]
```diff
@@ -0,0 +1,146 @@
+
```
[the remark this answers is not in the document]
Comment (MBkkt) [reply core/store/caching_directory.hpp:1]: You forgot to remove it

**Pull request merged**